Partial no-data GA mock reports

benbowler opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #10525

## Relevant technical choices

I added the `storybook/utils/` folder to the JS tests config.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.
- [ ] Ensure there are no unexpected significant changes to file sizes.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

zutigrm commented on storybook/utils/zeroReports.js:
```diff
@@ -39,6 +44,75 @@ export function replaceValuesInAnalytics4ReportWithZeroData( report ) {
+
+	return {
+		...report,
+		rows: removeDateRangeEntirely( rows ),
+		rowCount: removeDateRangeEntirely( rows ).length,
```
We can reduce repetition, and avoid double filtering here:

```js
const filteredRows = removeDateRangeEntirely( rows );
```
```suggestion
		rows: filteredRows,
		rowCount: filteredRows.length,
```

zutigrm commented on storybook/utils/zeroReports.js:
```diff
@@ -39,6 +44,75 @@ export function replaceValuesInAnalytics4ReportWithZeroData( report ) {
+
+		return {
+			...report,
+			rows: mapDateRangeToZero( rows ),
```
Do we want to calculate `rowCount` here as well?

zutigrm commented on storybook/utils/zeroReports.js:
```diff
@@ -39,6 +44,75 @@ export function replaceValuesInAnalytics4ReportWithZeroData( report ) {
+		const mapDateRangeToZero = ( cells ) =>
+			cells.map( ( cell ) => {
+				if (
+					cell.dimensionValues?.some(
+						( value ) => value.value === dateRangeKey
+					)
```
Since we use this same check in 2 places in this file, we can extract this into an util:
```js
const matchesDateRange = cell => cell.dimensionValues?.some( value => value.value === dateRangeKey );
```

And then use it here and bellow:
```suggestion
					matchesDateRange( cell )
```

zutigrm commented on storybook/utils/zeroReports.js:
```diff
@@ -39,6 +44,75 @@ export function replaceValuesInAnalytics4ReportWithZeroData( report ) {
+	// If emptyRowBehavior is 'remove', remove the rows and aggregate data that has a dimensionValues[].value equal to dateRangeKey.
+	const removeDateRangeEntirely = ( cells ) =>
+		cells.filter( ( cell ) => {
+			return ! cell.dimensionValues?.some(
+				( value ) => value.value === dateRangeKey
+			);
```
Also here:
```suggestion
			return ! matchesDateRange( cell )
```

zutigrm commented on storybook/utils/zeroReports.test.js:
```diff
@@ -0,0 +1,271 @@
+
+			// Check that date_range_0 rows have zeroed metric values
+			const dateRange0Rows = result.rows.filter( ( row ) =>
+				row.dimensionValues.some(
+					( value ) => value.value === 'date_range_0'
+				)
```
We have several repetitions of the same check/logic, that we can simplify by defining a util function also:
```js
const matchesDateRange = ( data, dateRange = 'date_range_0' ) => data.some( ( value ) => value.value === dateRange );
```

And then use it here and in other places in this file:
```suggestion
				matchesDateRange( row.dimensionValues )
```
And we would only need to pass additional parameters for date range where we check agains `date_range_1 `

zutigrm reviewed (changes requested):
Thanks @benbowler, nice work. I left you few comments

zutigrm commented on storybook/utils/zeroReports.test.js:
```diff
@@ -24,6 +24,9 @@ import mockedReportResponse from '../../assets/js/modules/analytics-4/utils/__fi
 
 describe( 'zeroReports', () => {
 	describe( 'replaceValuesOrRemoveRowForDateRangeInAnalyticsReport', () => {
+		const matchesDateRange = ( data, dateRange = 'date_range_0' ) =>
```
Since we are always passing ` dateRange`, we can remove default value. I will quickly push the change

zutigrm reviewed (approved):
Thanks @benbowler LGTM

tofumatt commented on storybook/utils/zeroReports.test.js:
```diff
@@ -0,0 +1,256 @@
+			expect( result.rows ).toHaveLength( 10 );
+
+			// Check that date_range_0 rows have zeroed metric values
+			const dateRange0Rows = result.rows.filter( ( row ) =>
```
Let's try to keep numbers out of variable names; I know there's nothing wrong with it but it's a bit uncommon is all. Plus: it makes them easier to differentiate/scan (`one` vs `zero` compared to `1` vs `0`).

Mostly though let's be more descriptive:

```suggestion
			const dateRangeWithZeroRows = result.rows.filter( ( row ) =>
```

tofumatt commented on storybook/utils/zeroReports.test.js:
```diff
@@ -0,0 +1,256 @@
+			} );
+
+			// Check that date_range_1 rows are unchanged
+			const dateRange1Rows = result.rows.filter( ( row ) =>
```
1 row isn't plural 😅 

```suggestion
			const dateRangeWithOneRow = result.rows.filter( ( row ) =>
```

tofumatt commented on storybook/utils/zeroReports.test.js:
```diff
@@ -0,0 +1,256 @@
+			} );
+
+			// Check that totals for date_range_0 are zeroed
+			const totalForDateRange0 = result.totals.find( ( total ) =>
```
```suggestion
			const totalForDateRangeWithZeroRowsInReport = result.totals.find( ( total ) =>
```

tofumatt commented on storybook/utils/zeroReports.test.js:
```diff
@@ -0,0 +1,256 @@
+			} );
+
+			// Check that totals for date_range_1 are unchanged
+			const totalForDateRange1 = result.totals.find( ( total ) =>
```
Okay, I'm going to leave my other suggestions in because it shows the variable names here are a bit unclear/ambiguous and confused me at first too. These are clearly keyed to "date range with the key `'date_range_0'`" and `"date range with the key `'date_range_1'`", but the variable names also read like ranges with a number of rows/results/data/etc.

"First date range" and "second date range", or "past date range" and "current date range" as descriptors would be clearer here, I think.

tofumatt reviewed (changes requested):
The variable names here could be cleaned up, but additionally: why do we have unit tests in the storybook utils folders?

I get it's because `zeroReports` is in Storybook, but these seem like useful tests outside Storybook, right? Should we move the utils to a most generic place? If not that's fine, just strikes me as a bit odd. 😅

benbowler commented:
Hey @tofumatt so I changed the variable names to use text numbers and removed some extra words that made them feel like they were referring to things other than `date_range_0` and `date_range_1`. These ranges are currently in the plugin the current and comparison date range but they don't have to be they could be many date ranges in the reports so I'd like to keep naming generic for future uses of the reporting infra.

I also moved the util to the js tests folder as it is used in js tests and stories and that means we don't have to run the main jest tests with files in the storybook folder. This does add quite a few file changes to update imports but I think it's worth it.
